File: firmware/app/commands.py
raw_commands = {
    'UP_FOR_LIMIT': 1,
    'STOP_AND_SET_TOP_LIMIT': 2,
    'GO_DOWN_FOR_LIMIT': 3,
    'STOP_AND_SET_BOTTOM_LIMIT': 4,
    'OPEN_BLINDS': 5,
    'CLOSE_BLINDS': 6,
    'MOVE_BLINDS_TO': 7,
    'READ_RESPONSE': 8,
    'PREPARE_STATUS': 9,
    'PREPARE_SETTINGS': 10,
    'UPDATE_SETTINGS': 11,
}
commands = {k: str(v) for k, v in raw_commands.items()}

from app.motor import Motor
from app.store import Store

class CommandProcesor:

  def __init__(self, is_left):
    self._store_0 = Store(0)
    self._store_1 = Store(1)
    self._m0 = Motor(is_left, self._store_0, {
      'in1': 9,
      'in2': 8,
      'pwm': 10,
      'stby': 16,
      'enc2': 21,
      'enc1': 20,
    })
    self._m1 = Motor(is_left, self._store_1, {
      'in1': 14,
      'in2': 15,
      'pwm': 12,
      'stby': 16,
      'enc2': 19,
      'enc1': 18,
    })

  def on_command(self, received_string):

    print(f'On command {received_string}')
    split_command = received_string.split(':')
    command_id = split_command[0]
    command_set = {k for k in commands if commands[k] == command_id}
    if len(command_set) == 1:
      command = list(command_set)[0]

      if command == 'PREPARE_SETTINGS':
        side = split_command[1]
        settings = self._store_0.get_settings_as_string() if side == '0' else self._store_1.get_settings_as_string()
        return settings
      if command == 'PREPARE_STATUS':
        side = split_command[1]
        return self._m0.format_status() if side == '0' else self._m1.format_status()  
      elif command == 'READ_RESPONSE':
        print('Reading response, doing nothing')
      elif command == 'UPDATE_SETTINGS':
        side = split_command[1] 
        store = self._store_0 if side == '0' else self._store_1
        store.store_settings_from_string(':'.join(split_command[2:]))
      elif command == 'UP_FOR_LIMIT':
        side = split_command[1]
        m = self._m0 if side == '0' else self._m1
        m.go_setup_up()
      elif command == 'STOP_AND_SET_TOP_LIMIT':
        side = split_command[1]
        m = self._m0 if side == '0' else self._m1
        m.stop_and_set_top_limit()
      elif command == 'GO_DOWN_FOR_LIMIT':
        side = split_command[1]
        m = self._m0 if side == '0' else self._m1
        m.go_setup_down()
      elif command == 'STOP_AND_SET_BOTTOM_LIMIT':
        side = split_command[1]
        m = self._m0 if side == '0' else self._m1
        m.finish_bottom_setup()
      elif command == 'MOVE_BLINDS_TO':
        side = split_command[1]
        m = self._m0 if side == '0' else self._m1
        m.go_to_position_percent(int(split_command[2]))                             
      else:  
        print(command)

    return ""

  def tick(self):
    self._m0.update()
    self._m1.update()

File: firmware/src/commands.cpp
#include "commands.h"


Commands::Commands(Motor* motor0, Motor* motor1, Store* store0, Store* store1) {
  _motor0 = motor0;
  _motor1 = motor1;
  _store0 = store0;
  _store1 = store1;
}

Motor* Commands::getMotor(uint8_t side) {
  return side == 0 ? _motor0 : _motor1;
}

Store* Commands::getStore(uint8_t side) {
  return side == 0 ? _store0 : _store1;
}

int Commands::onCommand(uint8_t* data, int length, uint8_t* nextPayload) {
  // return first 2 bytes to correlate ack payload in the receiver
  nextPayload[0] = data[0];
  nextPayload[1] = data[1];
  if (data[0] > static_cast<int>(Command::Command_MAX)) {
    Serial.printf("Command id '%d' is out of bounds\n", data[0]);
    return 2;
  }
  Command command = static_cast<Command>(data[0]);
  uint8_t side = data[1];
  if (command == Command::PREPARE_SETTINGS) {
    return getStore(side)->writeSettingsToBytes(&nextPayload[2]) + 2;
  } else if (command == Command::PREPARE_STATUS) {
    return getMotor(side)->writeStatus(&nextPayload[2]) + 2;
  } else if (command == Command::READ_RESPONSE) {
    // Serial.println("Reading response, doing nothing");
  } else if (command == Command::UPDATE_SETTINGS) {
    if (length != 17) {
      Serial.println("Updating settings failed, invalid size of the message");
      return 2;
    }
    getStore(side)->storeSettingsFromBytes(&(data[2]));
  } else if (command == Command::UP_FOR_LIMIT) {
    Serial.println("Processing UP_FOR_LIMIT");
    getMotor(side)->goSetupUp();
  } else if (command == Command::STOP_AND_SET_TOP_LIMIT) {
    getMotor(side)->stopAndSetTopLimit();
  } else if (command == Command::GO_DOWN_FOR_LIMIT) {
    getMotor(side)->goSetupDown();
  } else if (command == Command::STOP_AND_SET_BOTTOM_LIMIT) {
    getMotor(side)->stopAndFinishSetup();
  } else if (command == Command::MOVE_BLINDS_TO) {
    if (length != 3) {
      Serial.println("MOVE_BLINDS_TO failed, invalid size of the message");
      return 2;
    }
    getMotor(side)->goToPositionPercent(data[2]);
  } else if (command == Command::OPEN) {
    getMotor(side)->goToPositionPercent(0);
  } else if (command == Command::CLOSE) {
    getMotor(side)->goToPositionPercent(100);
  } else if (command == Command::RESET_BOTTOM) {
    getMotor(side)->resetBottom();
  } else {
    Serial.printf("Unknown command for id %d\n", data[0]);
  }

  return 2;
}

void Commands::tick() {
  _motor0->update();
  _motor1->update();
}